### ISSUE TITLE
Adapt to GTK theme

### ISSUE THREAD

**Issue by codic12** [akiraux/Akira#340]
I think that Akira has a lot of potential. However currently it only supports the elementary light/dark themes, not others. 
If this is possible and relatively easy, please add it - I'd be willing to contribute!

**Comment by Alecaddd** [comment]
Thanks for the comment.
Implementing other themes is currently out of scope and not a priority.
We will consider supporting other themes in the future.

**Comment by albfan** [comment]
@gsbhasin123 nothing prevents you from work on that but as @Alecaddd says is not a priority. 

Probably Akira will by distributed in sandbox (flatpak) so you will have an app exactly as we designed it.

**Comment by 12people** [comment]
It should be noted that neither GNOME nor Elementary supports GTK themes (which are just CSS overrides applied system-wide).

I'd be a fan of official support for Adwaita, as that's the only theme officially supported in GNOME.

To fully support the full possible range of CSS overrides would be an impossible task, and to target all the major themes out there would be a large effort that would require constant maintenance. I personally don't think that's quite worth the effort. There's a whole host of GTK developers that generally just target Adwaita, as attested to by the ["Please don't theme our apps" letter](https://stopthemingmy.app/).

**Comment by StarrKiss** [comment]
Why does this require targeting? I can work and change GTK themes to work on an app. I use the Breeze GTK theme, the default for KDE, and very widely used, but it shouldn't matter what theme I use. I use all sorts of GTK apps that conform to my desktop theme, and if an app doesn't support a theme, I complain to the theme devs. For example, GIMP doesn't work perfectly with every theme- why should I expect it to? I don't get this arbitary limitation. Just do what GIMP does, and BY DEFAULT use Adwaita packaged with the app, and if the USER wants they can click 'override' in the theme settings: 
![image](https://user-images.githubusercontent.com/13007697/89812105-393f7180-db05-11ea-8e3b-70b772f0eab9.png)
I don't get why apps do this. As to that 'open letter', I think a lot of it makes sense, and a lot of it is ridiculous. Nobody in their right minds expects the app DEV to support the theme, its the other way around. That's pretty obvious. But I don't see why the app developers should be allowed to override the user. A lot of use take pride and have fun in theming our desktops, and like having everything match. Why should our themes by put aside because the dev has a specific theme in mind for THEIR app? Just look how Akira clashes with every other app on my system-

![org](https://user-images.githubusercontent.com/13007697/89813130-ccc57200-db06-11ea-9964-c00044263ba4.png)

And even worse on the letter is icon themes. I don't get why its a shocker to devs that different people prefer different styles, and thats not bad! Who cares about "dev's intent" when in the end, its the USER using them? There are some people out there who like 3d icon themes and designs from the 90s. I may not be them, but I respect their choices. I just don't get the point of locking down the options of the user. At what point do we get to Microsoft, who don't allow ANY theming and have two options- Black and Blindingly white? The whole point of Linux is there's something for everyone.

**Comment by Alecaddd** [comment]
> But I don't see why the app developers should be allowed to override the user. A lot of use take pride and have fun in theming our desktops, and like having everything match. Why should our themes by put aside because the dev has a specific theme in mind for THEIR app? Just look how Akira clashes with every other app on my system-

This argument doesn't make sense at all and it shows how little you know about software development.

**Comment by StarrKiss** [comment]
@Alecaddd  I am not suggesting the dev supports the themes explicitly, I am merely suggesting you go the GIMP approach and allow GTK theming. Devs shouldn't have to support user provided themes, and users shouldn't have to use dev provided themes. And yes, I don't know that much about GTK, I write all my apps using QTQuick and C++.  On QT, you can force your own theme, too. But nobody actually does that. Instead, you can just let the user choose a QT theme, such as Kvantum or Breeze. I have encountered themes that break my QT apps (I write them to automate processes like bash scripts) and in that case I either don't use that theme or tweak my app. Case in point- recently, the Moe theme failed to work with certain apps, I contacted the DEVS of the moe theme and submitted a PR on GitLab.

**Comment by StarrKiss** [comment]
Also, please do not distribute this app in a sandbox or in FlatPak. I hate when devs make their app harder to tweak and use for no reason. Please just distribute it with Pacman, Apt, and DNF.

**Comment by bilelmoussaoui** [comment]
I think you are ignoring a lot of valid points against theming in general. It just a mess, you can't copy paste a bunch of CSS rules and expect every app to look the same (try to do that on your browser if you believe so....). The behaviour and design of the application is defined by the maker of the application not it's end users. Of course, you are free to contribute to it but that doesn't mean we most agree with what you think and do whatever you want just because it's open source.

Also Flatpak doesn't make app tweaking harder, it makes distributing software an easy process for the developers and for end users as they can grab an application and it works as the developer expected.

**Comment by Alecaddd** [comment]
> Also, please do not distribute this app in a sandbox or in FlatPak. I hate when devs make their app harder to tweak and use for no reason. Please just distribute it with Pacman, Apt, and DNF. 

Can you explain in details how would you tweak Akira after it has been released?

How is flatpak or snap harder to use? Can you give me some precise examples?

Akira is open source, and you're more than welcome to fork it, change it, and distribute it wherever you want.

**Comment by StarrKiss** [comment]
@bilelmoussaoui  Perhaps I am ignoring those points. I would agree on webpages, but with GUI toolkits, where everything is based on a set of widgets and toolkits, that's different. When writing QT apps, all the background and widgets based on QTQuick work fine with theming, and can all be tweaked at the same time. I've not noticed an issue there with any QT theme I try. And as I said, issues with themes (if they do have issues, I haven't encountered issues with themes in a VERY long time) should be reported to the theme maker.

> The behaviour and design of the application is defined by the maker of the application not it's end users

Why shouldn't users have freedom to tweak something beyond what the developer intended? Video games have had mods for decades now, including many that devs couldn't have intended, some of which are small and down to personal prefernece (think of the arachnophobia mods for games like Skyrim.) In the end, software should belong to the user, and that idea of software being user controlled is part of the open source idealogy in general. Everyone has different ideas, and everyone has different preferences. That's why theming EXIST! Some people like dark themes with yellow accents, other people like light themes with blur and transparency, some people like windows vista style 3d buttons, others like 90s pixel art buttons. A dev can't be expected to cater to all, hence- theming! I believe a dev has the obligation to allow, if not SUPPORT (in terms of issues) theming. As I said earlier- Doesn't that make us like Windows, having two themes that almost no apps follow? Perhaps I have underestimated the difficulty, but on the rare occaisions I write GTK apps with Glade, it uses the users theme BY DEFAULT. 
TL;DR for this section- software is written for the users- that's the whole point. Users should therefore have the right to tweak or change that software, EVEN IF IT MEANS overwriting the developers intentions. After all, is software made for the dev, or for the user?

>Also Flatpak doesn't make app tweaking harder, it makes distributing software an easy process for the developers and for end users as they can grab an application and it works as the developer expected

and a side effect to this is that each app is sandboxed, making it harder to interact with the rest of the system.

**Comment by StarrKiss** [comment]
> > Also, please do not distribute this app in a sandbox or in FlatPak. I hate when devs make their app harder to tweak and use for no reason. Please just distribute it with Pacman, Apt, and DNF.
> 
> Can you explain in details how would you tweak Akira after it has been released?

I would want to theme it and change the CSS on the fly to match my theme. That is my primary concern. 

> How is flatpak or snap harder to use? Can you give me some precise examples?

Snap and Flatpak sandbox apps, making it harder to interact with the system, AND ALSO MEANING they don't use system libraries, but instead use packaged versions of their own libraries. This has led to all sorts of issues in my experience, an example of which is an issue with the Snap version of Spotify not playing music properly on my Arch install due to conflicts with audio libraries, an issue I similarily had with the TextToSpeech in Discord snap. (Not to mention issues with proton and gfx drivers in FlatPak Steam, and the fact that Spicetify, a theming tool for Spotify, is BROKEN with snap: 
![image](https://user-images.githubusercontent.com/13007697/89826596-15d3f100-db1c-11ea-9e2c-944db47b229c.png)
)
Also I have had performance issues with Snap in the past, as well as the problems of the fact that SNAP stuff takes WAY longer to load due to it's sandboxing. 


> Akira is open source, and you're more than welcome to fork it, change it, and distribute it wherever you want.

Yes, but this is why I want theming and open resources, like the other poster mentioned. Not everyone is going to like my changes to Akira. This is not bad, it is mere taste. And say I modify Akira to allow theming, which clearly many people here dislike. (Again, a matter of taste). Now, say others make changes and Akira moves on, as things do. My patch might lack features as it is not in the main fork, so mine would become out of date. I would write such a patch right now if I knew enough about GTK. And a lot of other users might not know GTK , but still want to theme their computer. In the end, I don't see the reason for this artificial limiting of user control, esp. as GTK literally does the behavior I and themers want by default

**Comment by bilelmoussaoui** [comment]
> @bilelmoussaoui Perhaps I am ignoring those points. I would agree on webpages, but with GUI toolkits, where everything is based on a set of widgets and toolkits, that's different. When writing QT apps, all the background and widgets based on QTQuick work fine with theming, and can all be tweaked at the same time. I've not noticed an issue there with any QT theme I try. And as I said, issues with themes (if they do have issues, I haven't encountered issues with themes in a VERY long time) should be reported to the theme maker.

Web or almost every UI toolkit out there provides a set of basic widgets that you can extend by creating your custom ones. A widget is often created by either putting different widgets together or creating a custom one by drawing it yourself and adding the various interactions you want. IIRC exactly the same thing as in the web or Qt or whatever you want.

So you can't expect theme makers as you wish to call them, to track every GTK repository and look for every new custom widget to properly style it.That's just impossible. Not scalable. Even themes that are made by or "sponsored" by known distributions have failed and still failing to achieve that impossible task.

I don't think there are any reasons to continue this conversation and getting out of the topic.

**Comment by StarrKiss** [comment]
@bilelmoussaoui  Yes, but as I write custom QT widgets, I can write them (or, in fact, they are set by default) to inherit from theme colors officially supported in QT, such as background and highlight color. Say I write a rotating wheel widget. I can set the background of the wheel to inherit from widget_unselected_background, and the text of the wheel to inherit from widget_primary_text. Hence widgets can be made to work seamlessly with theming without issues or support from theme devs.

**Comment by bilelmoussaoui** [comment]
Well, how do you inherit colors in CSS? There's no public API that tells every theme should export all of those colors so that every different application with different needs will all of them have whatever they need.

**Comment by bilelmoussaoui** [comment]
Also open source doesn't mean you should be able to modify the app in fly heh. It means you can fork it and apply whatever changes you would like to see and re-compile or re-distribute it under the terms of the license. Which I suggest you do if you're unhappy with the current situation as I don't see that changing anytime soon.

**Comment by StarrKiss** [comment]
@bilelmoussaoui  I know nothing about GTK, so forgive me if this is wrong, but the theme I am using, breeze, has this gtk.css file:
![image](https://user-images.githubusercontent.com/13007697/89828092-2edda180-db1e-11ea-8704-2ebb3169126c.png)
Which seems very similar to QT to me in how colors are defined and apps read from them.

**Comment by StarrKiss** [comment]
@bilelmoussaoui  as to your second comment, I don't want to edit the apps code on the fly, I want to theme it, which means to apply custom CSS over the app. I can even do this on non- FOSS apps like Spotify:
![image](https://user-images.githubusercontent.com/13007697/89828225-62203080-db1e-11ea-813b-4ac62089b4c5.png)

**Comment by bilelmoussaoui** [comment]
Those colors are internal API. The apps don't have any access to them and differ from one them to another which makes it even more complex to base your application on something stable.

**Comment by StarrKiss** [comment]
@bilelmoussaoui  Well, I should do more research, but regardless, if the themes work, and the burden of support rests on the theme makers, why should an app dev NOT allow users to theme/modify the CSS of their app?

**Comment by bilelmoussaoui** [comment]
Because they are the app developers and are free to do whatever they want. As you said, the topic is more complex and i will suggest you do some research and play with things yourself before asking us to do something we technically can't fix, because it's not a technical issue. Not every issue can be fixed by "adding" or removing code.

**Comment by codic12** [comment]
This issue is evolving into a heated argument, definitely wasn't my intention. I see the developer's viewpoint, but still stand with my own. There must be some line where the theme is changed via a call to g_set_theme (or whatever it's called)... just change that and recompile if you want. Closing this issue for now. Very grateful to you all for creating and contributing to this amazing project

**Comment by albfan** [comment]
All @StarrKiss ranting is noise on this issue. We are open to work on support Adwaita theme, but right now it doesn't fit on our design. Theme is another dependency for our project, same is if you want us to use this or that dependency, simply we use whatever help us to reach our goal.

Complain about support other themes showing as example that KDE has its own, GTK has a default one and the user uses a different one shows that everybody attach to the theme they want for many different reasons.

flatpak is the app as we designed it, as long as you have the theme installed Akira works.

You want to tweak Akira? You're welcome, but don't try to force your vision on this (all apps should support my theme) blender, godot, KDE, are examples on the contrary. We are willing to see your solutions, not your complains.

"Adapt to GTK theme" is the topic here if we ever reopen this issue.
